src/crimson/common/buffer_seastar.cc

```diff
@@ -24,9 +26,22 @@ namespace ceph::buffer {
 
 class raw_seastar_foreign_ptr : public raw {
   seastar::foreign_ptr<temporary_buffer> ptr;
+  seastar::alien::instance& alien;
  public:
   raw_seastar_foreign_ptr(temporary_buffer&& buf)
-    : raw(buf.get_write(), buf.size()), ptr(std::move(buf)) {}
+    : raw(buf.get_write(), buf.size()), ptr(std::move(buf)),
+      alien(seastar::engine().alien()) {}
+
+  ~raw_seastar_foreign_ptr() {
+    if (ptr.get_owner_shard() != seastar::this_shard_id()) {
```

Comment:
I don't think seastar::this_shard_id() has a meaningful value on non-seastar threads.  From include/seastar/core/shard_id.hh:

```
using shard_id = unsigned;
...
#ifdef SEASTAR_BUILD_SHARED_LIBS
shard_id* this_shard_id_ptr() noexcept;
#else
inline shard_id* this_shard_id_ptr() noexcept {
    static thread_local shard_id g_this_shard_id;
    return &g_this_shard_id;
}
#endif
...
/// Returns shard_id of the of the current shard.
inline shard_id this_shard_id() noexcept {
    return *internal::this_shard_id_ptr();
}
```

from reactor.cc:

```
void smp::allocate_reactor(unsigned id, reactor_backend_selector rbs, reactor_config cfg) {
    assert(!reactor_holder);

    // we cannot just write "local_engin = new reactor" since reactor's constructor
    // uses local_engine
    void *buf;
    int r = posix_memalign(&buf, cache_line_size, sizeof(reactor));
    assert(r == 0);
    *internal::this_shard_id_ptr() = id;
    local_engine = new (buf) reactor(this->shared_from_this(), _alien, id, std::move(rbs), cfg);
    reactor_holder.reset(local_engine);
}
```

So, this_shard_id() dereferences this_shard_id_ptr(), which simply returns a pointer to an uninitialized unsigned value.  reactor.cc initializes this value for each reactor thread in allocate_reactor.  For a non-seastar thread, this_shard_id_ptr()::g_this_shard_id doesn't seem to be populated so seastar::this_shard_id will return unitialized memory which might, by accident, match a real seastar thread.  I think you need to check seastar::engine_is_ready() first.

Reply:
By looking at it again, we can simply remove this check. seastar::engine_is_ready() should be enough since it promises this condition already.